### PR TITLE
Issue 4746

### PR DESCRIPTION
**Pull request by guotong1988**
https://github.com/tensorflow/tensorflow/issues/4746

This code is the first commit to fix issue-4746 . I will commit the unit test code and doc later.
Thank you for your review . Thank you for any advice.I will work on it again for the advice and learn from the reviewers. 

I test this code with:

```
x_image = tf.placeholder(tf.float32,shape=[4,4])
x = tf.reshape(x_image,[1,4,4,1])

ksize_feed = np.array([1,2,2,1])

ksize_init = tf.placeholder(tf.int32, shape=[4])
ksize0 = tf.reshape(ksize_init,[2,2])
ksize1 = tf.reshape(ksize0,[1,2,2,1])

strides=[1,2,2,1]
padding = 'VALID'

y = tf.nn.max_pool(value=x,ksize=ksize1,strides=strides,padding=padding)

x_data = np.array([[4,3,1,8],[7,2,6,3],[2,0,1,1],[3,4,2,5]],
            dtype = np.float32)

with tf.Session() as sess:
    init = tf.initialize_all_variables()
    sess.run(init)

    x = (sess.run(x,feed_dict={x_image:x_data}))
    y = (sess.run(y,feed_dict={x_image:x_data,ksize_init:ksize_feed}))

    print "The shape of x:\t", x.shape, ",\t and the x.reshpe(4,4) is :"
    print x.reshape(4,4)
    print ""

    print "The shape of y:\t", y.shape, ",\t and the y.reshpe(2,2) is :"
    print y.reshape(2,2)
    print ""
```

### CODE REVIEW

**Comment by guotong1988** [comment]
Just test the code with:

```
import numpy as np
import tensorflow as tf

x_image = tf.placeholder(tf.float32,shape=[4,4])
x = tf.reshape(x_image,[1,4,4,1])

**ksize = np.array([1,2,2,1])**

strides=[1,2,2,1]
padding = 'VALID'

y = tf.nn.max_pool(value=x,ksize=ksize,strides=strides,padding=padding)

x_data = np.array([[4,3,1,8],[7,2,6,3],[2,0,1,1],[3,4,2,5]],
            dtype = np.float32)

with tf.Session() as sess:
    init = tf.initialize_all_variables()
    sess.run(init)

    x = (sess.run(x,feed_dict={x_image:x_data}))
    y = (sess.run(y,feed_dict={x_image:x_data}))

    print "The shape of x:\t", x.shape, ",\t and the x.reshpe(4,4) is :"
    print x.reshape(4,4)
    print ""

    print "The shape of y:\t", y.shape, ",\t and the y.reshpe(2,2) is :"
    print y.reshape(2,2)
    print ""
```

File: tensorflow/core/framework/common_shape_fns.cc [hunk 1]
```diff
@@ -448,7 +448,10 @@ Status AvgPoolShape(shape_inference::InferenceContext* c) {
 
 Status MaxPoolShape(shape_inference::InferenceContext* c) {
   ShapeHandle input_shape;
+  ShapeHandle kernel_sizes
```
Comment (drpngx) [inline tensorflow/core/framework/common_shape_fns.cc]: Did you try building this? This looks like a syntax error.

File: tensorflow/core/framework/common_shape_fns.cc [hunk 2]
```diff
@@ -481,16 +485,22 @@ Status MaxPoolShape(shape_inference::InferenceContext* c) {
     stride_depth = strides[1];
     stride_rows = strides[2];
     stride_cols = strides[3];
-    kernel_depth = kernel_sizes[1];
-    kernel_rows = kernel_sizes[2];
-    kernel_cols = kernel_sizes[3];
+    kernel_rows = c->Dim(kernel_sizes, 1);
+    kernel_cols = c->Dim(kernel_sizes, 2);
+    kernel_depth = c->Dim(kernel_sizes, 3);
+//    kernel_depth = kernel_sizes[1];
```
Comment (drpngx) [inline tensorflow/core/framework/common_shape_fns.cc]: We need to clean these up.

**Comment by zheng-xq** [comment]
Unfortunately, modifying an existing op like this breaks the backward compatiblity of TensorFlow. 

For now, it is a good idea to add a custom op on github outside TensorFlow, so people can try it if they are interested. When it gets popular and used in more models, we can move it to TF contrib or TF core. 

https://www.tensorflow.org/versions/r0.11/how_tos/adding_an_op/index.html

Thanks.

File: tensorflow/core/framework/common_shape_fns.cc (hunk 1 above)
Comment (guotong1988) [inline tensorflow/core/framework/common_shape_fns.cc]: It's my mistake.I did build it but I make a copy mistake.

**Comment by guotong1988** [comment]
OK , I will take another issue.

**Comment by drpngx** [comment]
Thanks for looking into this! I'll close the PR. Feel free to reopen if appropriate.

**Pull request closed without merging**